hunts/templates/all_puzzles.html

```diff
@@ -45,5 +46,276 @@ <h5 class="modal-title" id="addtitle">Add Puzzle</h5>
+        <small>
+            <span class="badge badge-${tag[1]}">${tag[0]}</span>
+            <button type="submit" class="btn btn-default" style="padding: 0px 0px;">
+                <i class="fa fa-xs fa-times"></i>
+            </button>
+        </small>
+    </form>`;
+                    }
+                    result += `
+<button type="button" class="addtag btn btn-default" data-toggle="modal" data-target="#addtag" data-pk="${row[pk]}" style="padding: 0px 0px;">
+    <i class="fa fa-xs fa-plus" aria-hidden="true"></i>
+</button>`
+                    return result;
+                }
+            },
+            {
+                'targets': metas,
+                'render': function(data, type, row, meta) {
+                    var result = `
+                    <button type="button" class="meta btn btn-outline-primary btn-sm" data-toggle="modal" data-target="#meta" data-pk="${row[pk]}">
+                        Assign Metas
+                    </button>`;
+
+                    return result;
+                }
+            }
+        ],
+    });
+
+    var all_uncollapsed = false;
+    function maybeExpandAll() {
+        if ( all_uncollapsed == false ) {
+            all_uncollapsed = true;
+            $('.tree').treegrid('expandAll');
+        }
+    }
 
+    function initTreegrid() {
+        $('.tree').treegrid({
+            onCollapse: function() {
+                all_uncollapsed = false;
+            }
+        });
+        if ($('.tree').hasClass( "initial-collapsed" )) {
+            $('.initial-collapsed').treegrid('collapse');
+        }
+    }
+
+    function areArraysEqual(arr1, arr2) {
+        return arr1.every((value, index) => {
+            if (Array.isArray(value)) {
+                return value.length === arr2[index].length
+                    && areArraysEqual(value, arr2[index]);
+            }
+            return value === arr2[index];
+        });
+    }
+
+    function getTableClassForPuzzleStatus(puzzleStatus) {
+        if (puzzleStatus === 'PENDING') {
+            return 'table-warning';
+        } else if (puzzleStatus === 'SOLVED') {
+            return 'table-success';
+        } else if (['STUCK', 'EXTRACTION'].includes(puzzleStatus)) {
+            return 'table-danger';
+        }
+        return '';
+    }
+
+    function updateRowColor(rowIndex, status) {
+        let newClass = getTableClassForPuzzleStatus(status);
+        let rowNode = table.row(rowIndex).node();
+        for (const tableClass of ['table-warning', 'table-danger', 'table-success']) {
+            $(rowNode).removeClass(tableClass);
+        }
+        $(rowNode).addClass(newClass);
+    }
+
+    function refreshModalHandlers() {
+        initEditPuzzle();
+        initDeletePuzzle();
+        initEditStatus();
+        initSubmitAnswer();
+        initAddTag();
+        initSetMeta();
+    }
+
+    function reload() {
+        $.ajax('/hunts/{{hunt_pk}}/puzzles', {
```

Comment:
This is going to make you lose your position in the table, isn't it? :(

Is there any way we could read that out and put you back to where you were?

Also, this is going to be annoying if you're in the middle of doing something and the page refreshes on you. I suggest doing the following:
- Track whether the user is idle (e.g. https://www.brcline.com/blog/detecting-inactivity-in-javascript or https://github.com/jasonmcleod/jquery.idle)
- When idle and the table is different, go ahead with the update
- If not idle, and the table is different, display an alert https://getbootstrap.com/docs/4.4/components/alerts/ with a button to update the table

Reply:
Ah, it actually won't make you lose your position because I use `table.draw(false)` which causes the current page position to be kept, even after redraw/reorder/research: https://datatables.net/reference/api/draw()

Reply:
I added a comment and switched to the equivalent `table.draw('full-hold')` for clarity:
```
                    // keep current page position after redraw
                    table.draw('full-hold');
```

Reply:
Also, the table is only redrawn when there are new puzzles added (`if (newdata.length !== olddata.length) { ....`), so most of the time, the refresh every 10 seconds won't do anything.

Reply:
When only data in a single row changes (e.g.: puzzle name or URL updated, puzzle status changes), only that row is invalidated and redrawn, and this shouldn't affect any search or paging the user has done.

Reply:
Awesome, thanks for thinking through the cases!